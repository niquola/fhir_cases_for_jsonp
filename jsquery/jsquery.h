/*-------------------------------------------------------------------------
 *
 * jsquery.h
 *     Definitions of jsquery datatype
 *
 * Copyright (c) 2014, PostgreSQL Global Development Group
 * Author: Teodor Sigaev <teodor@sigaev.ru>
 *
 * IDENTIFICATION
 *    contrib/jsquery/jsquery.h
 *
 *-------------------------------------------------------------------------
 */

#ifndef __JSQUERY_H__
#define __JSQUERY_H__

#include "fmgr.h"
#include "utils/numeric.h"
#include "utils/jsonb.h"

typedef struct
{
	int32	vl_len_;	/* varlena header (do not touch directly!) */
} JsQuery;

#define DatumGetJsQueryP(d)	((JsQuery*)DatumGetPointer(PG_DETOAST_DATUM(d)))
#define PG_GETARG_JSQUERY(x)	DatumGetJsQueryP(PG_GETARG_DATUM(x))
#define PG_RETURN_JSQUERY(p)	PG_RETURN_POINTER(p)


typedef struct JsQueryItem JsQueryItem;

struct JsQueryItem {
	enum {
		jqiNull = jbvNull, 
		jqiString = jbvString, 
		jqiNumeric = jbvNumeric, 
		jqiBool = jbvBool, 
		jqiArray = jbvArray, 
		jqiAnd = '&', 
		jqiOr = '|', 
		jqiNot = '!',
		jqiEqual = '=',
		jqiLess  = '<',
		jqiGreater  = '>',
		jqiLessOrEqual = '{',
		jqiGreaterOrEqual = '}',
		jqiContains = '@',
		jqiContained = '$',
		jqiOverlap = '%',
		jqiAny = '*',
		jqiAnyArray = '#',
		jqiKey = 'K',
		jqiIn = 'I'
	} type;

	union {
		struct {
			JsQueryItem	*left;
			JsQueryItem	*right;
		} args;

		JsQueryItem	*arg;

		Numeric		numeric;
		bool		boolean;
		struct {
			uint32      len;
			char        *val; /* could be not null-terminated */
		} string;

		struct {
			int			nelems;
			JsQueryItem	**elems;
		} array;
	};

	JsQueryItem	*next; /* next in path */
};

/*
 * support
 */

extern JsQueryItem* parsejsquery(const char *str, int len);

int32 readJsQueryHeader(char *base, int32 pos, int32 *type, int32 *nextPos);

#define read_byte(v, b, p) do {     \
	(v) = *(int8*)((b) + (p));      \
	(p) += 1;                       \
} while(0)                          \

#define read_int32(v, b, p) do {    \
	(v) = *(int32*)((b) + (p));     \
	(p) += sizeof(int32);           \
} while(0)                          \

#endif

