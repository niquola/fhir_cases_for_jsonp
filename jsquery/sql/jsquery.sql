CREATE EXTENSION jsquery;

set escape_string_warning=off;

select 'asd.zzz = 13'::jsquery;
select 'asd.zzz < 13'::jsquery;
select 'asd(zzz < 13)'::jsquery;
select 'asd(zzz < 13 & x = true)'::jsquery;
select 'asd(zzz < 13 & x = "true")'::jsquery;
select 'asd(zzz < 13 & x.zxc = "true")'::jsquery;
select 'asd(zzz < 13 | #.zxc = "true")'::jsquery;
select 'asd(* < 13 & #.zxc = "true")'::jsquery;
select '(* < 13 & #.zxc = "true")'::jsquery;
select '* < 13 & #.zxc = "true"'::jsquery;

select 'a < 1'::jsquery;
select 'a < -1'::jsquery;
select 'a < +1'::jsquery;
select 'a < .1'::jsquery;
select 'a < -.1'::jsquery;
select 'a < +.1'::jsquery;
select 'a < 0.1'::jsquery;
select 'a < -0.1'::jsquery;
select 'a < +0.1'::jsquery;
select 'a < 10.1'::jsquery;
select 'a < -10.1'::jsquery;
select 'a < +10.1'::jsquery;
select 'a < 1e1'::jsquery;
select 'a < -1e1'::jsquery;
select 'a < +1e1'::jsquery;
select 'a < .1e1'::jsquery;
select 'a < -.1e1'::jsquery;
select 'a < +.1e1'::jsquery;
select 'a < 0.1e1'::jsquery;
select 'a < -0.1e1'::jsquery;
select 'a < +0.1e1'::jsquery;
select 'a < 10.1e1'::jsquery;
select 'a < -10.1e1'::jsquery;
select 'a < +10.1e1'::jsquery;
select 'a < 1e-1'::jsquery;
select 'a < -1e-1'::jsquery;
select 'a < +1e-1'::jsquery;
select 'a < .1e-1'::jsquery;
select 'a < -.1e-1'::jsquery;
select 'a < +.1e-1'::jsquery;
select 'a < 0.1e-1'::jsquery;
select 'a < -0.1e-1'::jsquery;
select 'a < +0.1e-1'::jsquery;
select 'a < 10.1e-1'::jsquery;
select 'a < -10.1e-1'::jsquery;
select 'a < +10.1e-1'::jsquery;
select 'a < 1e+1'::jsquery;
select 'a < -1e+1'::jsquery;
select 'a < +1e+1'::jsquery;
select 'a < .1e+1'::jsquery;
select 'a < -.1e+1'::jsquery;
select 'a < +.1e+1'::jsquery;
select 'a < 0.1e+1'::jsquery;
select 'a < -0.1e+1'::jsquery;
select 'a < +0.1e+1'::jsquery;
select 'a < 10.1e+1'::jsquery;
select 'a < -10.1e+1'::jsquery;
select 'a < +10.1e+1'::jsquery;

select 'a in (0,1,2)'::jsquery;
select 'a in (0,null, "null", xxx, "zzz", 2)'::jsquery;

select '{"a": {"b": null}}'::jsonb @@ 'a.b = 1';
select '{"a": {"b": null}}'::jsonb @@ 'a.b = null';
select '{"a": {"b": null}}'::jsonb @@ 'a.b = false';
select '{"a": {"b": false}}'::jsonb @@ 'a.b = false';
select '{"a": {"b": false}}'::jsonb @@ 'a.b = true';
select '{"a": {"b": true}}'::jsonb @@ 'a.b = true';


select '{"a": {"b": 1}}'::jsonb @@ 'a.b = 1';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b < 1';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b <= 1';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b >= 1';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b > 1';

select '{"a": {"b": 1}}'::jsonb @@ 'a.b = 2';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b < 2';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b <= 2';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b >= 2';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b > 2';

select '{"a": {"b": 1}}'::jsonb @@ 'a.b = 0';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b < 0';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b <= 0';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b >= 0';
select '{"a": {"b": 1}}'::jsonb @@ 'a.b > 0';

select '{"a": {"b": 1}}'::jsonb @@ '*.b > 0';
select '{"a": {"b": 1}}'::jsonb @@ '*.b > 0';
select '{"a": {"b": 1}}'::jsonb @@ 'a.* > 0';
select '{"a": {"b": 1}}'::jsonb @@ 'a.* > 0';

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b && [ 1 ]';
select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b @> [ 1 ]';
select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b <@ [ 1 ]';
select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b @> [ 1,2,3,4 ]';
select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b <@ [ 1,2,3,4 ]';

select '[{"a": 2}, {"a": 3}]'::jsonb @@ '*.a = 4';
select '[{"a": 2}, {"a": 3}]'::jsonb @@ '*.a = 3';

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#.a = 4';
select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '*.a = 4';

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 1 | a=3)';
select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 3 | a=1)';
select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 3 & a=1)';
select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 3 & a=2)' as "false";
select '[{"a": 2, "b":3}, {"a": 3, "b": 1}]'::jsonb @@ '#(b = 1 & a=3)';

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#.a.a = 4';
select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '*.a.a = 4';
select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '*.#.a.a = 4';
select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#.*.a.a = 4';

select '{"a": 1}'::jsonb @@ 'a in (0,1,2)';
select '{"a": 1}'::jsonb @@ 'a in (0,2)';

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b.#=2';
select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b && [ 5 ]';

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a=*';
select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b=*';
select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.c=*';

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b = [1,2,3]';
select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b.# = [1,2,3]';
select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b && [1,2,3]';
select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b.# && [1,2,3]';

