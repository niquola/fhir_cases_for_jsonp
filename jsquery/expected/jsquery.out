CREATE EXTENSION jsquery;
set escape_string_warning=off;
select 'asd.zzz = 13'::jsquery;
     jsquery      
------------------
 "asd"."zzz" = 13
(1 row)

select 'asd.zzz < 13'::jsquery;
     jsquery      
------------------
 "asd"."zzz" < 13
(1 row)

select 'asd(zzz < 13)'::jsquery;
     jsquery      
------------------
 "asd"."zzz" < 13
(1 row)

select 'asd(zzz < 13 & x = true)'::jsquery;
            jsquery             
--------------------------------
 "asd"("zzz" < 13 & "x" = true)
(1 row)

select 'asd(zzz < 13 & x = "true")'::jsquery;
             jsquery              
----------------------------------
 "asd"("zzz" < 13 & "x" = "true")
(1 row)

select 'asd(zzz < 13 & x.zxc = "true")'::jsquery;
                jsquery                 
----------------------------------------
 "asd"("zzz" < 13 & "x"."zxc" = "true")
(1 row)

select 'asd(zzz < 13 | #.zxc = "true")'::jsquery;
               jsquery                
--------------------------------------
 "asd"("zzz" < 13 | #."zxc" = "true")
(1 row)

select 'asd(* < 13 & #.zxc = "true")'::jsquery;
             jsquery              
----------------------------------
 "asd"(* < 13 & #."zxc" = "true")
(1 row)

select '(* < 13 & #.zxc = "true")'::jsquery;
           jsquery           
-----------------------------
 (* < 13 & #."zxc" = "true")
(1 row)

select '* < 13 & #.zxc = "true"'::jsquery;
           jsquery           
-----------------------------
 (* < 13 & #."zxc" = "true")
(1 row)

select 'a < 1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < -1'::jsquery;
 jsquery  
----------
 "a" < -1
(1 row)

select 'a < +1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < .1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < -.1'::jsquery;
  jsquery   
------------
 "a" < -0.1
(1 row)

select 'a < +.1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < 0.1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < -0.1'::jsquery;
  jsquery   
------------
 "a" < -0.1
(1 row)

select 'a < +0.1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < 10.1'::jsquery;
  jsquery   
------------
 "a" < 10.1
(1 row)

select 'a < -10.1'::jsquery;
   jsquery   
-------------
 "a" < -10.1
(1 row)

select 'a < +10.1'::jsquery;
  jsquery   
------------
 "a" < 10.1
(1 row)

select 'a < 1e1'::jsquery;
 jsquery  
----------
 "a" < 10
(1 row)

select 'a < -1e1'::jsquery;
  jsquery  
-----------
 "a" < -10
(1 row)

select 'a < +1e1'::jsquery;
 jsquery  
----------
 "a" < 10
(1 row)

select 'a < .1e1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < -.1e1'::jsquery;
 jsquery  
----------
 "a" < -1
(1 row)

select 'a < +.1e1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < 0.1e1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < -0.1e1'::jsquery;
 jsquery  
----------
 "a" < -1
(1 row)

select 'a < +0.1e1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < 10.1e1'::jsquery;
  jsquery  
-----------
 "a" < 101
(1 row)

select 'a < -10.1e1'::jsquery;
  jsquery   
------------
 "a" < -101
(1 row)

select 'a < +10.1e1'::jsquery;
  jsquery  
-----------
 "a" < 101
(1 row)

select 'a < 1e-1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < -1e-1'::jsquery;
  jsquery   
------------
 "a" < -0.1
(1 row)

select 'a < +1e-1'::jsquery;
  jsquery  
-----------
 "a" < 0.1
(1 row)

select 'a < .1e-1'::jsquery;
  jsquery   
------------
 "a" < 0.01
(1 row)

select 'a < -.1e-1'::jsquery;
   jsquery   
-------------
 "a" < -0.01
(1 row)

select 'a < +.1e-1'::jsquery;
  jsquery   
------------
 "a" < 0.01
(1 row)

select 'a < 0.1e-1'::jsquery;
  jsquery   
------------
 "a" < 0.01
(1 row)

select 'a < -0.1e-1'::jsquery;
   jsquery   
-------------
 "a" < -0.01
(1 row)

select 'a < +0.1e-1'::jsquery;
  jsquery   
------------
 "a" < 0.01
(1 row)

select 'a < 10.1e-1'::jsquery;
  jsquery   
------------
 "a" < 1.01
(1 row)

select 'a < -10.1e-1'::jsquery;
   jsquery   
-------------
 "a" < -1.01
(1 row)

select 'a < +10.1e-1'::jsquery;
  jsquery   
------------
 "a" < 1.01
(1 row)

select 'a < 1e+1'::jsquery;
 jsquery  
----------
 "a" < 10
(1 row)

select 'a < -1e+1'::jsquery;
  jsquery  
-----------
 "a" < -10
(1 row)

select 'a < +1e+1'::jsquery;
 jsquery  
----------
 "a" < 10
(1 row)

select 'a < .1e+1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < -.1e+1'::jsquery;
 jsquery  
----------
 "a" < -1
(1 row)

select 'a < +.1e+1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < 0.1e+1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < -0.1e+1'::jsquery;
 jsquery  
----------
 "a" < -1
(1 row)

select 'a < +0.1e+1'::jsquery;
 jsquery 
---------
 "a" < 1
(1 row)

select 'a < 10.1e+1'::jsquery;
  jsquery  
-----------
 "a" < 101
(1 row)

select 'a < -10.1e+1'::jsquery;
  jsquery   
------------
 "a" < -101
(1 row)

select 'a < +10.1e+1'::jsquery;
  jsquery  
-----------
 "a" < 101
(1 row)

select '{"a": {"b": null}}'::jsonb @@ 'a.b = 1';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": null}}'::jsonb @@ 'a.b = null';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": null}}'::jsonb @@ 'a.b = false';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": false}}'::jsonb @@ 'a.b = false';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": false}}'::jsonb @@ 'a.b = true';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": true}}'::jsonb @@ 'a.b = true';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b = 1';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b < 1';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b <= 1';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b >= 1';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b > 1';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b = 2';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b < 2';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b <= 2';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b >= 2';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b > 2';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b = 0';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b < 0';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b <= 0';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b >= 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.b > 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ '*.b > 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ '*.b > 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.* > 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": 1}}'::jsonb @@ 'a.* > 0';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b && [ 1 ]';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b @> [ 1 ]';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b <@ [ 1 ]';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b @> [ 1,2,3,4 ]';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b <@ [ 1,2,3,4 ]';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}]'::jsonb @@ '*.a = 4';
 ?column? 
----------
 f
(1 row)

select '[{"a": 2}, {"a": 3}]'::jsonb @@ '*.a = 3';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#.a = 4';
 ?column? 
----------
 f
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '*.a = 4';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 1 | a=3)';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 3 | a=1)';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 3 & a=1)';
 ?column? 
----------
 f
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#(a = 3 & a=2)' as "false";
 false 
-------
 f
(1 row)

select '[{"a": 2, "b":3}, {"a": 3, "b": 1}]'::jsonb @@ '#(b = 1 & a=3)';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#.a.a = 4';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '*.a.a = 4';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '*.#.a.a = 4';
 ?column? 
----------
 t
(1 row)

select '[{"a": 2}, {"a": 3}, {"a": {"a":4}}]'::jsonb @@ '#.*.a.a = 4';
 ?column? 
----------
 t
(1 row)

select '{"a": 1}'::jsonb @@ 'a && [0,1,2]';
 ?column? 
----------
 t
(1 row)

select '{"a": 1}'::jsonb @@ 'a && [0,2]';
 ?column? 
----------
 f
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ 'a.b.#=2';
 ?column? 
----------
 t
(1 row)

select '{"a": {"b": [1,2,3]}}'::jsonb @@ '*.b && [ 5 ]';
 ?column? 
----------
 f
(1 row)

